src/DIRAC/FrameworkSystem/private/authorization/utils/Tokens.py

```diff
@@ -0,0 +1,227 @@
+import time
+import json
+import datetime
+
+from DIRAC import S_OK, S_ERROR
+from DIRAC.Core.Utilities import DErrno
+from DIRAC.ConfigurationSystem.Client.Helpers import Registry
+from DIRAC.Resources.IdProvider.IdProviderFactory import IdProviderFactory
+
+from authlib.oauth2.rfc6749.util import scope_to_list
+from authlib.oauth2.rfc6749.wrappers import OAuth2Token as _OAuth2Token
+from authlib.integrations.sqla_oauth2 import OAuth2TokenMixin
+
+
+def getTokenLocation():
+  """ Research token file location. Use the bearer token discovery protocol
+      defined by the WLCG (https://zenodo.org/record/3937438) to find one.
+
+      :return: str
+  """
+  if os.environ.get('BEARER_TOKEN_FILE'):
+    return os.environ['BEARER_TOKEN_FILE']
+  elif os.environ.get('XDG_RUNTIME_DIR'):
+    return "%s/bt_u%s" % (os.environ['XDG_RUNTIME_DIR'], os.getuid())
+  else:
+    return "/tmp/bt_u%s" % os.getuid()
+
+
+def getLocalTokenDict(location=None):
+  """ Search local token. Use the bearer token discovery protocol
+      defined by the WLCG (https://zenodo.org/record/3937438) to find one.
+
+      :param str location: environ variable name or file path
+
+      :return: S_OK(dict)/S_ERROR()
+  """
+  env = (location if location and location.startswith('/') else None) or 'BEARER_TOKEN'
+  if os.environ.get(env):
+    return S_OK(OAuth2Token(os.environ[env]))
+  return readTokenFromFile(location if location and location.startswith('/') else None)
+
+
+def readTokenFromFile(fileName=None):
+  """ Read token from a file
+
+      :param str fileName: filename to read
+
+      :return: S_OK(dict)/S_ERROR()
+  """
+  location = fileName or getTokenLocation()
+  try:
+    with open(location, 'rt') as f:
+      token = f.read()
+  except IOError as e:
+    return S_ERROR(DErrno.EOF, "Can't open %s token file.\n%s" % (location, repr(e)))
+  return S_OK(OAuth2Token(token))
+
+
+def writeToTokenFile(tokenContents, fileName):
+  """ Write a token string to file
+
+      :param str tokenContents: token as string
+      :param str fileName: filename to dump to
+
+      :return: S_OK(str)/S_ERROR()
+  """
+  location = fileName or getTokenLocation()
+  try:
+    with open(location, 'wt') as fd:
+      fd.write(tokenContents)
+  except Exception as e:
+    return S_ERROR(DErrno.EWF, " %s: %s" % (location, repr(e)))
+  try:
+    os.chmod(location, stat.S_IRUSR | stat.S_IWUSR)
+  except Exception as e:
+    return S_ERROR(DErrno.ESPF, "%s: %s" % (location, repr(e)))
+  return S_OK(location)
+
+
+def writeTokenDictToTokenFile(tokenDict, fileName=None):
+  """ Write a token dict to file
+
+      :param dict tokenDict: dict object to dump to file
+      :param str fileName: filename to dump to
+
+      :return: S_OK(str)/S_ERROR()
+  """
+  fileName = fileName or getTokenLocation()
+  if not isinstance(tokenDict, dict):
+    return S_ERROR('Token is not a dictionary')
+  return writeToTokenFile(json.dumps(tokenDict), fileName)
+
+
+class OAuth2Token(_OAuth2Token):
```

Comment:
The WLCG Bearer Token Discovery document also mentions handling multiple tokens, for cases where multiple VOs are being used. I know very little about multi-vo DIRAC but is this something that should be included in the design of this interface?

Reply:
It is not implemented yet, but it can be implemented later based on practical use